apis/apiextensions/v1/composition_environment.go

```diff
@@ -0,0 +1,217 @@
+/*
+Copyright 2022 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package v1
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
+)
+
+const (
+	errFmtInvalidEnvironmentSourceSelectorLabelMatcherType = "invalid label matcher type '%s'"
+)
+
+// An EnvironmentConfiguration specifies the environment for rendering composed
+// resources.
+type EnvironmentConfiguration struct {
+	// EnvironmentConfigs selects a list of `EnvironmentConfig`s. The resolved
+	// resources are stored in the composite resource at
+	// `spec.environmentConfigRefs` and is only updated if it is null.
+	//
+	// The list of references is used to compute an in-memory environment at
+	// compose time. The data of all object is merged in the order they are
+	// listed, meaning the values of EnvironmentConfigs with a larger index take
+	// priority over ones with smaller indices.
+	//
+	// The computed environment can be accessed in a composition using
+	// `FromEnvironmentFieldPath` and `CombineFromEnvironment` patches.
+	// +optional
+	EnvironmentConfigs []EnvironmentSource `json:"environmentConfigs,omitempty"`
+
+	// Patches is a list of environment patches that are executed before a
+	// composition's resources are composed.
+	Patches []EnvironmentPatch `json:"patches,omitempty"`
+}
+
+// EnvironmentSourceType specifies the way the EnvironmentConfig is selected.
+type EnvironmentSourceType string
+
+const (
+	// EnvironmentSourceTypeReference by name.
+	EnvironmentSourceTypeReference EnvironmentSourceType = "Reference"
+	// EnvironmentSourceTypeSelector by labels.
+	EnvironmentSourceTypeSelector EnvironmentSourceType = "Selector"
+)
+
+// EnvironmentSource selects a EnvironmentConfig resource.
+type EnvironmentSource struct {
+	// Type specifies the way the EnvironmentConfig is selected.
+	// Default is `Reference`
+	// +optional
+	// +kubebuilder:validation:Enum=Reference;Selector
+	// +kubebuilder:default=Reference
+	Type EnvironmentSourceType `json:"type,omitempty"`
+
+	// Ref is a named reference to a single EnvironmentConfig.
+	// Either Ref or Selector is required.
+	// +optional
+	Ref *EnvironmentSourceReference `json:"ref,omitempty"`
```

Comment:
I left it out, because it is implied implicitly by lying inside `EnvironmentSource`:

```yaml
spec:
  environment:
    environmentConfigs:
    - type: Reference
      reference:
        name: example-environment
```